### ISSUE TITLE
Support in spillover workflow for "BCKG190Di" column header

### ISSUE THREAD

**Issue by jmw86069** [HelenaLC/CATALYST#35]
## Scenario

The Compensation spillover workflow fails for the column header "**BCKG190Di**".
Specifically, `computeSpillmat()` interprets any column containing numeric
values to be an isotope, and "BCKG190Di" is interpreted as element "BCKG"
with mass 190. Later `plotSpillmat()` calls `CATALYST:::check_sm()` which
fails to find "BCKG" in `names(isotope_list)`.

More generally, `plotSpillmat()` (actually `CATALYST::check_sm()`)
produces an error  when a column header containing a numeric digit does not
represent an element found in `names(isotope_list)`.

## Proposed resolutions

There are three workarounds I suggest to the authors, in order of
potential effort and hopefully positive impact to CATALYST:

### 1. **Add "BCKG" to isotope_list**

In this case, add "BCKG" with one mass value 190, to `isotope_list`,
representing the column header "**BCKG190Di**". This workaround is the
lowest-effort option, and enables one additional Fluidigm output format to
work with CATALYST.

### 2. **`computeSpillmat()` should validate elements against `isotope_list`**

This validates the data before the downstream `plotSpillmat()` fails.
It can then choose how to handle missing entries gracefully.
I thought of two options, but I defer to the authors:

* ***Elements not found in `names(isotope_list)` are ignored***, the mass
  value is changed to NA, which will remove them from downstream
  processing.
* ***Elements not found in `names(isotope_list)` are added to
  `isotope_list`*** with one mass value derived from the column header,
  e.g. "**BCKG190Di**" would add an entry, for example:
  `isotope_list <- c(isotope_list, list(BCKG=c(190))`.

I am unclear whether entries such as "**BCKG190Di**" should be included or
excluded.

### 3. **Allow user-defined `isotope_list`**.

This workaround is perhaps the most effort, but might be preferred
for long-term support of CATALYST. It relieves the authors from downstream
support of the `isotope_list` values, and empowers users to self-support
their workflows. It also allows analysts to customize the `isotope_list`
entries if that mechanism might be useful.

The current code directly uses `CATALYST::isotope_list` however, that
mechanism could be changed. One option is to create a new
function `get_isotope_list()` which retrieves the appropriate `isotope_list`
either from `.GlobalEnv` or from `CATALYST::isotope_list`. By this mechanism,
the CATALYST shinyGui would need a mechanism to use a custom isotope_list.
One simple possibility is to load isotope_list into the user R environment
before running `CATALYST::launchGUI()`.

## Next steps

I have forked the repository to test option #1 above, but will defer
to the authors about how to proceed. I'm happy to help contribute!
Thank you for your assistance!

**Comment by jmw86069** [comment]
I really appreciate how quickly you implemented those enhancements!
I **tested and confirmed** the changes you made resolved the error in
our workflow.

The only remaining/new issue to allow users of the shinyGui to provide or
use a custom `isotope_list`. I can make this request a separate issue if
you prefer?

My earlier thought was that updating `CATALYST::isotope_list` at the package
level would have also resolved the shiny app using a custom `isotope_list`,
but only for my specific needs. However, another user-supportable option
is to allow the user running `CATALYST::launchGUI()` to supply a custom
`isotope_list`.

In the spirit of requesting a feature alongside potential options, I have
a couple ideas. :-) Feel free to discard if you have better ideas in mind!
Basically my suggestion is not to implement custom `isotope_list` at the
shinyGui level, but at the step of starting the shinyGui. That lets the user
self-support without adding complexity to the web GUI.

If supplying a parameter to `CATALYST::launchGUI` works, it would be
straightforward:

```R
isotope_list2 <- c(CATALYST::isotope_list, list(BCKG=153));
CATALYST::launchGUI(isotope_list=isotope_list2);
```

My experience with shiny has not used that path, so I'm guessing it
does not work.

Alternative is to supply `"isotope_list.txt"` in the working directory
when `CATALYST::launchGUI()` is run. If that file exists, it gets used
(with some verbose message indicating as such). There are benefits to
allowing text files, mainly they're user-supportable, visible, easy to
debug, etc. Similar to the typical linux-style config files. In essence,
if a custom `isotope_list` is supplied as a file, I suggest a text file
rather than .rdata/.rds.

One potentially easy file format could be a two-column `data.frame`:

Element | Mass
--- | ---
Ge | 70
Ge | 72
Ge | 73
Ge | 74
Ge | 76
AS | 75
BCKG | 190
... | ...

(One could create the data.frame like this:
```R
data(isotope_list);
isotope_list_df <- data.frame(
   element=rep(names(isotope_list), lengths(isotope_list)),
   mass=unlist(isotope_list)
   )
## Add one row
isotope_list_df <- rbind(isotope_list_df,
   data.frame(element="BCKG", mass=190));
## Save a custom isotope_list.txt
write.table(x=isotope_list_df,
   file="isotope_list.txt",
   sep="\t");
```
)

Then `CATALYST::launchGUI()` would import the data.frame, then split into a
list:

```R
if (file.exists("isotope_list.txt")) {
   isotope_list_df <- read.table("isotope_list.txt", sep="\t");
   isotope_list <- split(isotope_list_df[,2], isotope_list_df[,1]);
} else {
   isotope_list <- CATALYST::isotope_list;
}
```

There are a bunch of possible implementations, whatever works is fine with
me! I just wanted to make sure my suggestion had supporting R code to show
what I was thinking.

**Comment by jmw86069** [comment]
Per your question about `computeSpillmat()` not including `isotope_list`
as an optional parameter, I agree I don't see the need to add it there.

It ignores colnames like `c("Residual","Width")` since they don't contain
numeric values. However, if some file contains a header like `"Offset50"`
then I'm not sure what it would do. I suppose it would run the calculations
then simply be ignored during the `plotSpillmat()` step? If so, then it
would not pose a problem.

**Comment by HelenaLC** [comment]
**Regarding a custom isotope_list for Shiny**
I've implemented this as a "mix" of your 2 suggestions (see server.R):
- On start-up, we check `exists("custom_isotope_list")`, which will be used if it exists. Else, we default to `CATALYST::isotope_list`. A message notifying the user of whether the "default" or "custom" isotope list is being used will be shown in the bottom right corner. 
- Specifically, this requires the user to specify a variable `custom_isotope_list` in the environment prior to running `launchGUI()`/running the app with a button push. 
- The `custom_isotope_list` is created *de novo* or from "default" as you suggested earlier, e.g. `custom_isotope_list <- list(CATALYST::isotope_list, BCKG=190)`
- I chose this solution as
  - novice/online users won't be affected by it
  - no extra file has to be created/put into the wd
- Lastly, the specified custom list will be checked on startup (see server.R). If any of the checks fails, the user will be notified of this and the app will again go with the default list.

---

**Regarding `computeSpillmat()`**
Yes, indeed this will be ignored in the calculations, unless (!) 
1. 50 is in `bc_ms` when deconvolution the single-stains, AND
2. Offset = c(49,50,51,56,xxx) (for example) is in the `isotope_list`

This is because
1. If the mass isn't specified when calling `assignPrelim()`, no singly positive population will be identified and no spill can be estimated later on.
2. `computeSpillmat()` runs a helper function `CATALYST:::get_spill_cols` which will get the +/-1M (49,51), +16M (56), isotopes (xxx) channels and estimate spill there. 

**In summary:** If anyone knows what they're doing this should not happen and a channel such as "Offset50" will not pose a problem, especially if the numeric value (here 50) is out of the range of the masses of interest.

**Comment by jmw86069** [comment]
Thank you again for your updates and comments!

I performed a quick test and it gave an error indicating that `plotSpillmat()`
did not recognize the argument `isotope_list`, which appears to be from
one of the alternate signatures in the `plotSpillmat.R` file. I should say,
my "quick test" involved creating `custom_isotope_list` but slightly
different syntax:

```R
custom_isotope_list <- c(CATALYST::isotope_list, list(BCKG=190))
```

The step "Estimate spillover" is what caused the error, shown below:

```diff
- Error: unused argument (isotope_list = list(Ge = c(70, 72, 73, 74, 76), 
- As = 75, Se = c(74, 76, 77, 78, 80, 82), Br = c(79, 81), Kr = c(78, 80, 
- 82, 83, 84, 86), Rb = c(85, 87), Sr = c(84, 86, 87, 88), Y = 89, Zr = 
- c(90, 91, 92, 94, 96), Nb = 93, Mo = c(92, 94, 95, 96, 97, 98, 100), 
- Tc = c(97, 98, 99), Ru = c(96, 98, 99, 100, 101, 102, 104), Rh = 103, 
- Pd = c(102, 104, 105, 106, 108, 110), Ag = c(107, 109), Cd = c(106, 108, 
- 110, 111, 112, 113, 114, 116), In = c(113, 115), Sn = c(112, 114, 115, 
- 116, 117, 118, 119, 120, 122, 124), Sb = c(121, 123), Te = c(120, 122, 
- 123, 124, 125, 126, 128, 130), I = 127, Xe = c(124, 126, 128, 129, 130, 
- 131, 132, 134, 136), Cs = 133, Ba = c(130, 132, 134, 135, 136, 137, 138), 
- La = c(138, 139), Ce = c(136, 138, 140, 142), Pr = 141, Nd = c(142, 143, 
- 144, 145, 146, 148, 150), Pm = c(145, 147), Sm = c(144, 147, 148, 149, 
- 150, 152, 154), Eu = c(151, 153), Gd = c(152, 154, 155, 156, 157, 158, 
- 160), Tb = 159, Dy = c(156, 158, 160, 161, 162, 163, 164), Ho = 165, 
- Er = c(162, 164, 166, 167, 168, 170), Tm = 169, Yb = c(168, 170, 171, 
- 172, 173, 174, 176), Lu = c(175, 176), Hf = c(174, 176, 177, 178, 179, 
- 180), Ta = c(180, 181), W = c(180, 182, 183, 184, 186), Re = c(185, 187), 
- Os = c(184, 186, 187, 188, 189, 190, 192), Ir = c(191, 193), Pt = c(190, 
- 192, 194, 195, 196, 198), Au = 197, Hg = c(196, 198, 199, 200, 201, 202, 
- 204), Tl = c(203, 205), Pb = c(204, 206, 207, 208), Bi = 209, BCKG = 190))
```

It looks like a relatively simple edit to that file will fix the issue,
to add `isotope_list` to the definition for the two alternate `plotSpillmat()`
functions, then adding `isotope_list=isotope_list` to their calls to the
default `plotSpillmat()` function.

Come to think of it, should the function calls be ignoring the user-defined
parameters? I could be misinterpreting the syntax, but noticed it when looking
at the plotSpillmat.R file.

Instead of:

```R
# ------------------------------------------------------------------------------
#' @rdname plotSpillmat
setMethod(f="plotSpillmat",
    signature=signature(bc_ms="ANY", SM="data.frame"),
    definition=function(bc_ms, SM, 
        out_path=NULL, name_ext=NULL, annotate=TRUE, plotly=TRUE) {
        plotSpillmat(bc_ms, as.matrix(SM),
            out_path=NULL, 
            name_ext=NULL,
            annotate=TRUE,
            plotly=TRUE)
    }
)
```

It would become:

```R
# ------------------------------------------------------------------------------
#' @rdname plotSpillmat
setMethod(f="plotSpillmat",
    signature=signature(bc_ms="ANY", SM="data.frame"),
    definition=function(bc_ms, SM,
        isotope_list=CATALYST::isotope_list,
        out_path=NULL, name_ext=NULL, annotate=TRUE, plotly=TRUE) {
        plotSpillmat(bc_ms, as.matrix(SM),
            out_path=out_path, 
            isotope_list=isotope_list,
            name_ext=name_ext,
            annotate=annotate,
            plotly=plotly)
    }
)
```

**Comment by jmw86069** [comment]
Just an FYI, I did some testing, updating my fork of your CATALYST package, and noticed that the `compCytof()` generics also need to have `isotope_list` added as parameters. After making that update, it worked successfully!

In my fork, I made one other small change to the `differential_analysis.Rmd` vignette to check for `diffcyt` package before proceeding with the vignette, since I'm using R-3.4 and cannot install `diffcyt` yet. I'm happy to defer to your R changes for long term use, or submit my changes as a pull request, no worries either way.

Thank you again for this package and your help!
